app/v2/controllers/V2MovementsController.scala

```diff
@@ -314,6 +319,34 @@ class V2MovementsControllerImpl @Inject() (
         }
     }
 
+  private def acceptedMessageBody(
+    messageSummary: MessageSummary,
+    acceptHeader: String,
+    bodyAndSizeMaybe: Option[BodyAndSize]
+  )(implicit
+    hc: HeaderCarrier
+  ): EitherT[Future, PresentationError, BodyAndContentType] = {
+    def bodyExists(bodyAndSize: BodyAndSize): EitherT[Future, PresentationError, BodyAndContentType] =
+      acceptHeader match {
+        case VersionedRouting.VERSION_2_ACCEPT_HEADER_VALUE_JSON if bodyAndSize.size > config.smallMessageSizeLimit =>
+          EitherT.leftT[Future, BodyAndContentType](PresentationError.notAcceptableError("Large messages cannot be returned as json"))
+        case VersionedRouting.VERSION_2_ACCEPT_HEADER_VALUE_JSON =>
+          for {
+            jsonStream <- conversionService.xmlToJson(messageSummary.messageType.get, bodyAndSize.body).asPresentation
+            bodyWithContentType = BodyAndContentType(MimeTypes.JSON, jsonStream)
+          } yield bodyWithContentType
+        case VersionedRouting.VERSION_2_ACCEPT_HEADER_VALUE_XML =>
+          EitherT.rightT[Future, PresentationError](
+            BodyAndContentType(MimeTypes.XML, bodyAndSize.body)
+          )
+      }
+    bodyAndSizeMaybe match {
+      case Some(value) => bodyExists(value)
+      case None =>
+        EitherT.leftT[Future, BodyAndContentType](PresentationError.notFoundError(s"Body for message id ${messageSummary.id} does not exist"))
+    }
```

Comment:
Slight nitpicky but good practice -- can we leave a new line between the function and this? It took me a moment to properly comprehend where the function ended.

I'm a fan of the inner function itself though.